src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Core/Usage/CollectionPropertiesShouldBeReadOnly.cs

```diff
@@ -0,0 +1,113 @@
+    /// for an example of this pattern.
+    /// 
+    /// Both binary and XML serialization support read-only properties that are collections.
+    /// The System.Xml.Serialization.XmlSerializer class has specific requirements for types that implement ICollection and 
+    /// System.Collections.IEnumerable in order to be serializable.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class CollectionPropertiesShouldBeReadOnlyAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2227";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.CollectionPropertiesShouldBeReadOnly), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.CollectionPropertiesShouldBeReadOnlyMessage), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                         s_localizableTitle,
+                                                                         s_localizableMessage,
+                                                                         DiagnosticCategory.Usage,
+                                                                         DiagnosticSeverity.Warning,
+                                                                         isEnabledByDefault: true,
+                                                                         helpLinkUri: "https://msdn.microsoft.com/library/ms182327.aspx",
+                                                                         customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(Rule);
+            }
+        }
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.RegisterCompilationStartAction(
+                (context) =>
+                {
+                    var iCollectionType = context.Compilation.GetICollectionType();
+                    var arrayType = context.Compilation.GetArrayType();
+
+                    if (iCollectionType == null || arrayType == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(iCollectionType, arrayType);
```

Comment:
Do you prefer explicitly creating the class here? You could simply have a lambda that captures these two variables and the compiler will generate the class. In general, I make the class explicit only when there's a lot of state being tracked and we are doing multiple things with that object. I'd suggest writing this as 
```csharp
context.RegisterSymbolAction(symbolContext => AnalyzeSymbol(iCollectionType, arrayType), SymbolKind.Property);
```

and make `AnalyzeSymbol` a static method on this class.